collaboration/extending-kyma-cli.md

```diff
@@ -0,0 +1,351 @@
+
+```bash
+  <function-name>
+  ├── .vscode
+  │   └── launch.json
+  ├── deployment
+  │   └── deployment.yaml
+  ├── local
+  │   └── <server-code.extension>
+  └── Readme.md
+  └── code
+    ├── config.yaml
+    ├── <function-code.extension>
+    └── <dependecies.extension>
+```
+
+* launch.json consists of settings for the local debugger on vscode
+* deployment.yaml consists of the yaml of the function cr that has been deployed on the kyma cluster. This would be generated once function has been deployed.
+* `<server-code.extension>` consists of the server code with handler pointing to <function-code.extension>. The `extension` of file and `server-code` is with repsect to the language passed in init command
+* config.yaml can be used as info file with current configurations like:
+  * name
+  * namespace
+  * events subscribed
+  * environment variables
+  * Api name to expose with
+  * replicas
+  * min/max memory
+  * min/max cpu
+  * runtime
+  * debug command (to be used with telepresence. This can help if we want to have different runtimes)
+* <function-code.extension> where the developer can write his logic to test. The `extension` is language specific.
+* `dependecies.extension` file with list of dependencies. eg. `package.json` for node and `go.mod and go.sum` for golang dependencies
+
+>The templates generated should be placed a directory in the library for the functions. We should have templates for each language that can be supported. In the future if we are supporting more runtime we can move it to separate repo.
+
+##### config.yaml
+
+This file basically enables developers to configure their functions. With init command this file is created and it basically has some default values. The readme files provide information about various options present in this file
+
+Developer can modify this file to deploy or update a function on to the kyma cluster. This would enable the developer not only tune the function (by modifying replica, memory cpu), but also provide an overview of the current triggers for the function.
+
+Usually one does not change the config so often during development compared to code. So our aim should be to make deployment and testing the function simpler. So having a config file taking care of configuration would make deployment much simpler.
+
+Having such a config file would help developers with limited or no kubernetes knowledge to get themselves aquainted with various configuration options for function and concentrate on the business logic for writing the functions.
+
+Additionally it can be used automated deployments of functions. As this file can be read by kyma cli to deloy function with required configurations.
+
+Example structure of config.yaml
+
+```yaml
+name: <function_name>
+namespace: <namespace> #default by default
+events:
+  - event1
+  - event2
+environment-vars:
+  - foo:bar
+api:
```

Comment:
So it will also define APIRules. Sounds nice, but am afraid to support here the full feature set, think of all the different auth options. So either we limit it to a very specific minimal feature set or we have a generic mechanism in place which mainly takes the content here and aplies it 1:1 to the apiserver. What is the plan?

Reply:
APIRules would be implemented as a library. The feature set would be defined what has been implemented inside the library.

Reply:
(1) If we implement something like this should we give the user the possibility to output the yamls which would be used to realize the function defined in `config.yaml`. Same as with `deployment/function.yaml` ?
(2) I like the idea, but that sounds like another API for functions (and also eventing, authentication etc.). What about we translate this config file or some kyma cli invocation (e.g. ```kyma function deploy --namespace <namespace> --name <name> --trigger-event stage.com.external.solution.order.created ...```) to the actual kubernetes CRs and put them inside the `deployment` folder ?  So not using a `config.yaml` at all.